// ionic/commands/capacitor/copy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const color_1 = require("../../lib/color");
const base_1 = require("./base");
class CopyCommand extends base_1.CapacitorCommand {
    async getMetadata() {
        return {
            name: 'copy',
            type: 'project',
            summary: 'Copy web assets to native platforms',
            description: `
${color_1.input('ionic capacitor copy')} will do the following:
- Copy the ${color_1.strong('www/')} directory into your native platforms.
      `,
            inputs: [
                {
                    name: 'platform',
                    summary: `The platform to copy (e.g. ${['android', 'ios', 'electron'].map(v => color_1.input(v)).join(', ')})`,
                },
            ],
        };
    }
    async preRun(inputs, options, runinfo) {
        await this.preRunChecks(runinfo);
        if (inputs[0]) {
            await this.checkForPlatformInstallation(inputs[0]);
        }
    }
    async run(inputs, options) {
        const [platform] = inputs;
        const args = ['copy'];
        if (platform) {
            args.push(platform);
        }
        await this.runCapacitor(args);
    }
}
exports.CopyCommand = CopyCommand;
