// ionic/commands/capacitor/sync.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const color_1 = require("../../lib/color");
const base_1 = require("./base");
class SyncCommand extends base_1.CapacitorCommand {
    async getMetadata() {
        return {
            name: 'sync',
            type: 'project',
            summary: 'Sync (copy + update) an Ionic project',
            description: `
${color_1.input('ionic capacitor sync')} will do the following:
- Copy web assets to all Capacitor native platforms
- Update each Capacitor native platforms, such as any dependencies that need updating.
- Install any discovered Capacitor or Cordova plugins.
      `,
            inputs: [
                {
                    name: 'platform',
                    summary: `The platform to sync (e.g. ${['android', 'ios', 'electron'].map(v => color_1.input(v)).join(', ')})`,
                },
            ],
        };
    }
    async preRun(inputs, options, runinfo) {
        await this.preRunChecks(runinfo);
        if (inputs[0]) {
            await this.checkForPlatformInstallation(inputs[0]);
        }
    }
    async run(inputs, options) {
        const [platform] = inputs;
        const args = ['sync'];
        if (platform) {
            args.push(platform);
        }
        await this.runCapacitor(args);
    }
}
exports.SyncCommand = SyncCommand;
